src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/References/ReferenceCleanupService.ReferenceHandlers.cs

```diff
@@ -0,0 +1,185 @@
+                var references = await GetReferencesAsync(selectedConfiguredProject, cancellationToken);
+
+                ProjectSystemReferenceInfo referenceInfo = references.FirstOrDefault(c => c.ItemSpecification == referenceUpdate.ReferenceInfo.ItemSpecification);
+
+                if (referenceInfo is null)
+                {
+                    return false;
+                }
+
+                return referenceInfo.TreatAsUsed != true;
+            }
+
+            internal async Task<List<ProjectSystemReferenceInfo>> GetReferencesAsync(ConfiguredProject selectedConfiguredProject, CancellationToken cancellationToken)
+            {
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var references = new List<ProjectSystemReferenceInfo>();
+
+                var projectItems = await GetUnresolvedReferencesAsync(selectedConfiguredProject);
+
+                if (projectItems is null)
+                {
+                    return references;
+                }
+
+                foreach (var item in projectItems)
+                {
+                    string? treatAsUsed = await GetAttributeTreatAsUsedAsync(item.Metadata);
+                    string? itemSpecification = item.UnevaluatedInclude;
+
+                    if (!string.IsNullOrEmpty(itemSpecification))
+                    {
+                        references.Add(new ProjectSystemReferenceInfo(_referenceType, itemSpecification, treatAsUsed == bool.TrueString));
+                    }
+                }
+
+                return references;
+            }
+
+            private static async Task<string?> GetAttributeTreatAsUsedAsync(IProjectProperties metadata)
+            {
+                string? value = await metadata.GetUnevaluatedPropertyValueAsync(ProjectReference.TreatAsUsedProperty);
+                return value ?? bool.FalseString;
+            }
+
+            internal async Task<bool> UpdateReferenceAsync(ConfiguredProject selectedConfiguredProject, ProjectSystemReferenceUpdate referenceUpdate, CancellationToken cancellationToken)
+            {
+                bool wasUpdated = false;
+
+                cancellationToken.ThrowIfCancellationRequested();
+
+                var projectItems = await GetUnresolvedReferencesAsync(selectedConfiguredProject);
+
+                var item = projectItems
+                    .FirstOrDefault(c => c.UnevaluatedInclude == referenceUpdate.ReferenceInfo.ItemSpecification);
+
+                if (item != null)
+                {
+                    string newValue = referenceUpdate.Action == ProjectSystemUpdateAction.TreatAsUsed ? bool.TrueString : bool.FalseString;
+
+                    await item.Metadata.SetPropertyValueAsync(ProjectReference.TreatAsUsedProperty, newValue, null);
+
+                    wasUpdated = true;
+                }
+
+                return wasUpdated;
+            }
+        }
+
+        private class ProjectReferenceHandler : AbstractReferenceHandler
```

Comment:
We place nested classes in their own file.

Reply:
Done.
Moved AbstractReferenceHandler to its own file.

I wanted to keep this class private, so only ReferenceCleanupService will use it.